### PR TITLE
Add xoauth2 support for imap and smtp

### PR DESCRIPTION
**Pull request by ryanakca**
* **What does this PR do?**

It adds XOAUTH2 authentication support for IMAP and SMTP. XOAUTH2 is an authentication scheme used by OWA/Office365, and is required by some organizations.

* **Does this PR meet the acceptance criteria?** (This is just a reminder for you,
  this section can be removed if you fulfill it.)

   - Documentation created/updated (you have to edit
     [doc/manual.xml.head](https://www.github.com/neomutt/neomutt/blob/master/doc/manual.xml.head)
     for that)

Yes.

   - All builds and tests are passing

Unrelated tests failed. (symlink, chmod, etc.)

   - Added [doxygen code documentation](https://neomutt.org/dev/doxygen)
     [syntax](http://www.stack.nl/~dimitri/doxygen/manual/docblocks.html)

Yes.

   - Code follows the [style guide](https://neomutt.org/dev/coding-style)

As best as I could.

* **What are the relevant issue numbers?**

https://github.com/neomutt/neomutt/issues/2927

### CODE REVIEW

File: docs/manual.xml.head [hunk 1]
```diff
@@ -12073,8 +12073,25 @@ set imap_oauth_refresh_command="/path/to/oauth2.py --quiet --user=[email_address
 </screen>
 
       <para>
-        Substitute pop or smtp for imap in the above example to configure for
-        those.
+        For Office365, you can use the <literal>mutt_oauth2.py</literal> script
+        written by Alexander Perlis:
+        <ulink url="https://gitlab.com/muttmua/mutt/-/blob/master/contrib/mutt_oauth2.py">https://gitlab.com/muttmua/mutt/-/blob/master/contrib/mutt_oauth2.py</ulink>
```
Comment (gahr) [inline docs/manual.xml.head:12078]: Do we perhaps want to add these two files to our contrib directory?

File: send/smtp.c [hunk 2]
```diff
@@ -576,6 +576,40 @@ static int smtp_auth_oauth(struct SmtpAccountData *adata, const char *method)
...
+  FREE(&oauthbearer);
+  FREE(&ibuf);
+
+  if (rc == -1)
+    return SMTP_AUTH_FAIL;
+  if (smtp_get_resp(adata) != 0)
+    return SMTP_AUTH_FAIL;
+
+  return SMTP_AUTH_SUCCESS;
+}
```
Comment (gahr) [inline send/smtp.c:611]: How about refactor `smtp_auth_oauth` and `smtp_auth_xoauth2` into a common function?

```c
static int smtp_auth_oauth_xoauth2(struct SmtpAccountData *adata, const char *method, bool xoauth2)
{
  (void) method; // This is XOAUTH2

  // L10N: (%s) is the method name, e.g. Anonymous, CRAM-MD5, GSSAPI, SASL
  mutt_message(_("Authenticating (%s)..."), method);

  /* We get the access token from the smtp_oauth_refresh_command */
  char *oauthbearer = mutt_account_getoauthbearer(&adata->conn->account, xoauth2);
  if (!oauthbearer)
    return SMTP_AUTH_FAIL;

  size_t ilen = strlen(oauthbearer) + 30;
  char *ibuf = mutt_mem_malloc(ilen);
  snprintf(ibuf, ilen, "AUTH %s %s\r\n", method, oauthbearer);

  int rc = mutt_socket_send(adata->conn, ibuf);
  FREE(&oauthbearer);
  FREE(&ibuf);

  if (rc == -1)
    return SMTP_AUTH_FAIL;
  if (smtp_get_resp(adata) != 0)
    return SMTP_AUTH_FAIL;

  return SMTP_AUTH_SUCCESS;
}

static int smtp_auth_oauth(struct SmtpAccountData *adata, const char *method)
{
  return smtp_auth_oauth_xoauth2(adata, method, false);
}

static int smtp_auth_xoauth2(struct SmtpAccountData *adata, const char *method)
{
  return smtp_auth_oauth_xoauth2(adata, method, true);
}
```

**Comment by gahr** [comment]
Thanks, this looks neat!

File: docs/manual.xml.head (hunk 1 above)
Comment (ryanakca) [reply docs/manual.xml.head]: Imported them!

File: send/smtp.c (hunk 2 above)
Comment (ryanakca) [reply send/smtp.c:601]: Done. Refactored imap authentication as well.

**Review by flatcap — COMMENTED** [review]
I've moved the contrib scripts into a separate dir, and altered the Makefile to install them.
(I adjusted the docs to match)

Plus, a few minor tweaks:
- Rebased over master
- Minor cosmetic (clang-format, etc)

**Review by gahr — APPROVED** [review]

**Comment by flatcap** [comment]
Sorry about the delay, I've been on holiday.
Merged.  Thank you very much!

**Pull request merged**